### ISSUE TITLE
Create a published `testkit` module

### ISSUE THREAD

**Issue by armanbilge** [http4s/http4s#5320]
cc @amesgen. Linking to:
https://github.com/http4s/http4s-jdk-http-client/issues/398
https://github.com/http4s/http4s/discussions/4140#discussioncomment-265556

Now that we have two satellite modules `http4s-jdk-http-client` and `http4s-dom` and with the ever impending great schism this is only all the more important. 

AFAICT mainly `ClientRouteTestBattery` has been discussed ... I think it would be valuable to publish a testkit module for _both_ clients and servers. As I understand it, a "TestBattery" is a (`HttpApp`, `Suite`) pair (where a suite is a collection of tests) , with a pluggable server and pluggable client. Depending on what you want to test, ideally one of those should be a "control" 😁 so it seems to me without much additional work, a solid battery could be used for both?

Besides a battery, I wonder what else might be useful in a `testkit` module?

**Comment by rossabaker** [comment]
Yes, I would imagine both a client and server version.  I suppose it would generally need to be blackbox: everything over a TCP socket.  Also is the suite small for the minimal things every backend must do, or large to expose what they can and cannot do?  The behavior will never be 100% aligned.
 
I think this is great and needed, but, whew, it's daunting.

**Comment by armanbilge** [comment]
Thank you both for the feedback 😅 yes this is  _very_ daunting and I hope we can start small by publishing the `ClientRouteTestBattery` and the server scaffolds so that we can use them in `jdk-http-client` and the `FetchClient` in `http4s-dom`. Hope that @amesgen can weigh in too when they have a chance.

> In it's current shape, the assertions in these tests look into blaze, which makes the tests impossible to reuse for ember.

Yes, I did see that. I was initially thinking we could just abstract an interface by which the client backend could the necessary information (I'm not sure how hard it would be to implement this for ember). But, if this could be done with a special test server that sounds cool idea, tests relying on sniffing sounds complicated but I wouldn't know 😅

**Comment by armanbilge** [comment]
The `ServerScaffold` introduced in the following two PRs can be the start of our test kit, it just needs to be extracted out of the `client` module into a new `client-testkit` module, if anyone would like to work on that? :)
- https://github.com/http4s/http4s/pull/5601
- https://github.com/http4s/http4s/pull/5587

**Comment by armanbilge** [comment]
This issue blocks schisming of the various client backends.
